### PR TITLE
Process Video Locally Infragram.org

### LINKED ISSUE #418: GSoC Infragram.org full-screen UI and video upload Discussion and Planning
Hi, everyone! This is the Planning Issue for the [Infragram.org: full-screen UI and video upload](https://summerofcode.withgoogle.com/programs/2022/projects/rP8hthBg) for [infragram.org](https://infragram.org/). The project is in collaboration with the [Outreachy proposal: Design a new full-screen user interface for Infragram.org](https://publiclab.org/notes/stefq1111/04-08-2022/outreachy-proposal) . All are open for discussion and improvement.

**Process Video Locally**
- [x] Allow conversion of entire video.
- [x] Implement video controls on canvas.
- [x] Include another Canvas Element ( second Canvas ).
- [x] Record  video on process at second canvas.
- [x]  Save processed video file.
- [x] Integrate the above functionalities to with UI of outreachy collaborator. @stephaniequintana 

**Allow selecting multiple output resolutions as well**
- [x] Dynamically change the size of second canvas to fit required resolution.
- [x] Integrate the above functionalities to with UI of outreachy collaborator.

**Integrate Teamsys latest camera selection API** 
 Issue #39
- [ ] Download Temasys/AdapterJS library.
- [ ] Change all instances of the word `webRtcOptions.context ` to `webRtcDetectedType`
- [ ] Remove code `if (webRtcOptions.context === "flash")`
[Reference 1](https://github.com/publiclab/spectral-workbench.js/pull/172)  [Reference 2]( https://github.com/publiclab/spectral-workbench.js/issues/87)

**Accept multiple webcam resolutions**
- [x] Maintain css canvas size.
- [x] Change HTML canvas size to fit requested resolution.
- [x] Integrate the above functionalities to with UI of outreachy collaborator.

**Write UI tests for new Infragram UI Jest** 
- [ ] write test for infragram UI functionalities.

**Allow zooming/panning of video**
- [x]  Zooming :  Add uScale of uniform vect2 to shader.frag and shader.vert. 
- [ ]  Panning :  Add uPan of uniform vect2 to shader.frag and shader.vert .
- [ ] Implement above functionalities to infragram drawscreen function.
- [ ] Integrate the above functionalities to with UI of outreachy collaborator.

**Miscellaneous**

- [ ]  Add option to close webcam input
- [ ]  Enable image and video stream mirroring
- [x]  Add file filters on input to optimize infrared analysis

### PR DESCRIPTION
**Pull request by Forchapeatl**
**Process Video Locally**  
Fixes #418  
found at [https://forchapeatl.github.io/infragram/index2.html](https://forchapeatl.github.io/infragram/index2.html)

https://user-images.githubusercontent.com/24577149/174498132-3f0b65a8-a498-40e8-8fd3-6c8f0295fa56.mp4

- [x] Allow conversion of entire video.
- [x] Implement video controls on canvas.

**TEST RESULTS**

**Passed Test cases**
- video can be sleeked through muted / looped / paused and played.
- video can be processed.
- Video processing works before and after camera has been on.

**Failed Test cases**
- Accommodate  large video files > 10MB.

Hello @jywarren @TildaDares @stephaniequintana
The demo  video is processed locally , but it might take a while for the video to load up the [github page](https://forchapeatl.github.io/infragram/index2.html)
It seems to work only after camera has been turned on.This bug is taking a while.

- I plead you give me a week with the bug while I proceed to other mile stones. 
- Is the code base readable (in terms of variable declarations)

Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

* [x] PR is descriptively titled 📑 and links the original issue above 🔗
* [x] tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR
* [x] code is in uniquely-named feature branch and has no merge conflicts 📁
* [x] screenshots/GIFs are attached 📎 in case of UI updation
* [x] ask `@publiclab/reviewers` for help, in a comment below

### CODE REVIEW

**Comment by jywarren** [comment]
Hi! I hope to have a chance to give this a thorough read through in the next day, but I wonder if the camera start bug has to do with having to wait until it's initialized; there may be a need to listen for the right event?

File: index3.html [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+        trigger: "hover",
+      });
+      var infragram, openInPl, openInSequencer;
+      (function() {
+        infragram = Infragram({
+          uploader: false,
+          processor: 'webgl'
+        });
+
+        openInPl = function openInPl() {
```
Comment (jywarren) [inline index3.html:261]: I guess we should move this into a shareable .js file or into the library itself... Just an idea?

File: index3.html [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+            <p>OUTPUT</p>
+            <button onClick="openInPl()" type="button" class="btn btn-primary btn-lg">Save</button>
+            <button onClick="infragram.download()" type="button" class="btn btn-primary btn-lg">Download</button>
+            <button onClick="openInSequencer()" type="button" class="btn btn-primary btn-lg">Export</button>
+          </div>
+
+          <br style="clear:both;" />
+
+        </div>
+        <div id="localVideoControls" style="display:none;">
```
Comment (jywarren) [inline index3.html:202]: So let's clearly label and explain these to @stephaniequintana who may be soon integrating this code into the new UI!

File: dist/infragram3.js [hunk 3]
```diff
@@ -0,0 +1,2213 @@
+ isVideo = false,isCamera=false; // Turns off camera feed
```
Comment (jywarren) [inline dist/infragram3.js:1]: Ah ok! Did you make changes to a copy of dist/infragram.js? So what may be better is to look where in /src/ you want to make the changes, and do them there... Then they'll generate a new dist/infragram.js when we run `grunt build`. Does that make sense? It compiles the code into /dist/. @cesswairimu or @TildaDares both have experience with this too! No worries we'll help if it doesn't make sense.

File: index3.html (hunk 1 above)
Comment (Forchapeatl) [reply index3.html:261]: Please help me with more details on this file. I can't seem to find a file named sharable.js in the infragram directory , I am somehow confused on the mentioned library. Please  some detail on this idea.

File: dist/infragram3.js (hunk 3 above)
Comment (Forchapeatl) [reply dist/infragram3.js:1]: yes the `grunt build` command makes sense, thank you.

File: index3.html (hunk 1 above)
Comment (jywarren) [reply index3.html:261]: Sorry I'm typing from phone, I just meant a separate js file we can include into the different html files... A shareable file. Thanks!

Comment (Forchapeatl) [reply index3.html:261]: Okay. Noted

Comment (Forchapeatl) [reply index3.html:261]: thank you

**Comment by jywarren** [comment]
Hi @Forchapeatl - i wanted to try to help you pull apart the new code you've added to the `/dist/infragram.js` file, and so I made a new PR just to get a good 'diff' view -- at https://github.com/publiclab/infragram/pull/422/files?diff=unified&w=1

What I'm going to try to do is comment on each section and point to where those changes could be made in the /src/ directory. Let's try to make the changes there!

**Comment by jywarren** [comment]
OK - it's not too complex - there are about 6 or so sections which need to be copied into corresponding code in `/src/io/camera.js` and `/src/ui/interface.js` -- press "load diff" and you'll see my comments at https://github.com/publiclab/infragram/pull/422/files?diff=unified&w=1 !

If you make those changes in the `Forchapeatl-patch-1` branch, then run `grunt build`, you'll generate a new version of the compiled file called `dist/infragram.js`. Check that file in and then remove `/dist/infragram3.js`, and point `index3.html` at `/dist/infragram.js`, and we should be OK!

**Comment by Forchapeatl** [comment]
okay @jywarren ,thank you

**Comment by Forchapeatl** [comment]
Hello  @jywarren , @cesswairimu , @TildaDares  the code base has been structured as requested. on https://github.com/publiclab/infragram/pull/422/files?diff=unified&w=1

**Comment by jywarren** [comment]
This looks excellent @Forchapeatl -- can you run `grunt build` and check in the updated `dist/....` files? Then I'm going to give it a quick test in GitPod and we should be able to merge -- great work!!

**Comment by jywarren** [comment]
No rush as it's very late here so I won't get to it until tomorrow. Congrats on the changes, they were complex!

**Comment by Forchapeatl** [comment]
Thank you, @jywarren

**Comment by jywarren** [comment]
Hi @Forchapeatl I didn't see the dist files added again, just wanted to check if you needed any help?

**Comment by Forchapeatl** [comment]
Hello @jywarren , all is well with grunt.  The functionality is a bit buggy . Please I will  include the dist file once I am done with these bugs. 
>Video processing works only after camera has been on.
>Accommodate large video files > 10MB.

Permit me redirect you to another PR [https://github.com/publiclab/infragram/pull/427](https://github.com/publiclab/infragram/pull/427)

**Comment by jywarren** [comment]
> > Video processing works only after camera has been on.
> > Accommodate large video files > 10MB.

No problem, but I think the 10MB might be a tough one - i'm not sure how you're storing it but if it's in localStorage, there is a hard JS limit on how much can be stored. If you think it's a big issue, we could push people into the workflow of recording on their device, then uploading the recording and converting it, rather than recording live within the website. What do you think? So people would upload a video file in the same way as they would have uploaded a still image. @stephaniequintana also curious what you think about that workflow as an alternative. It could be simpler and could sidestep some of these issues?

**Comment by jywarren** [comment]
I want to say that we have processed larger images in https://github.com/publiclab/image-sequencer/ or https://github.com/jywarren/webgl-distort/ - I could be wrong, but I think we were just keeping such larger amounts of data in memory, or in the HTML DOM itself. Where is the 10mb limit happening?

File: src/ui/interface.js [hunk 4]
```diff
@@ -96,6 +103,61 @@ module.exports = function Interface(options) {
...
+      }             
+      isCamera = false;
+      if(!isVideo){ //Prevent Creation of Duplicate video Elements
+        localVideo = document.createElement('video');
+        localVideo.setAttribute("src", "../video.mp4");
+        localVideo.setAttribute("id", "localVideo");
+        localVideo.style.display = "none"               
+        localVideo.style.width = "100px";   
+        localVideo.style.height = "50px";    
+        document.getElementById("video-container").appendChild(localVideo);
```
Comment (jywarren) [inline src/ui/interface.js:118]: It looks like everything is happening in the DOM here. Did you get a particular error you can share?

Comment (Forchapeatl) [reply src/ui/interface.js:118]: https://github.com/publiclab/infragram/blob/Forchapeatl-patch-1/src/Infragram.js#L38-L39


Sorry about the mix up @jywarren . I meant that HD videos have a slower frame rate transition on the Infragram canvas.

For example,  HD [video](https://github.com/bower-media-samples/big-buck-bunny-1080p-30s/blob/master/video.mp4) makes better frame transitions with the below settings
```
    if (options.processor.type == "webgl") interval = 1;
    else interval = 15;
```

Comment (jywarren) [reply src/ui/interface.js:118]: I see, ok! What did you think about making this driven by frame conversion completion instead of interval?

**Comment by Forchapeatl** [comment]
Video Processing works now works  without clicking the webcam button. 


https://user-images.githubusercontent.com/24577149/179400204-050add7e-c809-48ed-a3a5-8a9e9fddd425.mp4

**Comment by Forchapeatl** [comment]
You 're welcome  @jywarren

**Comment by jywarren** [comment]
Ok no worries and great to see the other pr! My internet isn't great here
so I hope to review tomorrow. If you can get some feedback from another
mentor or a fellow intern before then that's great too!

**Pull request closed without merging**